File: src/db/collection.py
"""
VideoCollection container for rich query results, aggregations, and transformations.
"""

from typing import (
    List,
    Iterator,
    Optional,
    Callable,
    Union,
    Dict,
    Any,
    Set,
    Sequence,
    overload,
)
import json
import random
from pathlib import Path

from .models import Video, Language, Speaker


class VideoCollection(Sequence[Video]):
    """
    An immutable, rich collection of Wikitongues Video records with chainable
    transformations, aggregations, stats, and export capabilities.
    """

    def __init__(self, videos: Optional[Sequence[Video]] = None):
        self._videos: List[Video] = list(videos) if videos is not None else []

    def __len__(self) -> int:
        return len(self._videos)

    def __iter__(self) -> Iterator[Video]:
        return iter(self._videos)

    @overload
    def __getitem__(self, index: int) -> Video:
        ...

    @overload
    def __getitem__(self, index: slice) -> "VideoCollection":
        ...

    def __getitem__(self, index: Union[int, slice]) -> Union[Video, "VideoCollection"]:
        if isinstance(index, slice):
            return VideoCollection(self._videos[index])
        return self._videos[index]

    def __contains__(self, item: Any) -> bool:
        if isinstance(item, Video):
            return item in self._videos
        if isinstance(item, str):
            # Allow checking containment by video ID
            return any(v.id == item for v in self._videos)
        return False

    def __repr__(self) -> str:
        lang_count = len(self.iso_codes)
        return f"<VideoCollection: {len(self._videos)} video(s) across {lang_count} language(s)>"

    # -------------------------------------------------------------------------
    # Aggregation & Summary Properties
    # -------------------------------------------------------------------------

    @property
    def is_empty(self) -> bool:
        return len(self._videos) == 0

    @property
    def ids(self) -> List[str]:
        """List of YouTube video IDs."""
        return [v.id for v in self._videos]

    @property
    def urls(self) -> List[str]:
        """List of YouTube video URLs."""
        return [v.url for v in self._videos]

    @property
    def embed_urls(self) -> List[str]:
        """List of YouTube embed player URLs."""
        return [v.embed_url for v in self._videos]

    @property
    def titles(self) -> List[str]:
        """List of video titles."""
        return [v.title for v in self._videos]

    @property
    def total_duration_seconds(self) -> int:
        """Total duration across all videos in seconds."""
        return sum(v.duration_seconds for v in self._videos)

    @property
    def total_duration_formatted(self) -> str:
        """Human-readable total duration formatted as 'Xh Ym Zs'."""
        total = self.total_duration_seconds
        hours = total // 3600
        minutes = (total % 3600) // 60
        seconds = total % 60
        parts = []
        if hours > 0:
            parts.append(f"{hours}h")
        if minutes > 0 or hours > 0:
            parts.append(f"{minutes}m")
        parts.append(f"{seconds}s")
        return " ".join(parts)

    @property
    def average_duration_seconds(self) -> float:
        """Average duration per video in seconds."""
        if not self._videos:
            return 0.0
        return self.total_duration_seconds / len(self._videos)

    @property
    def iso_codes(self) -> List[str]:
        """Distinct primary ISO 639-3 codes in the collection."""
        seen: Set[str] = set()
        codes: List[str] = []
        for v in self._videos:
            code = v.primary_language.iso639_3
            if code and code not in seen:
                seen.add(code)
                codes.append(code)
        return codes

    @property
    def bcp47_codes(self) -> List[str]:
        """Distinct primary BCP 47 codes in the collection."""
        seen: Set[str] = set()
        codes: List[str] = []
        for v in self._videos:
            bcp = v.primary_language.bcp47
            if bcp and bcp not in seen:
                seen.add(bcp)
                codes.append(bcp)
        return codes

    @property
    def language_names(self) -> List[str]:
        """Distinct primary language names in the collection."""
        seen: Set[str] = set()
        names: List[str] = []
        for v in self._videos:
            name = v.primary_language.name
            if name and name not in seen:
                seen.add(name)
                names.append(name)
        return names

    @property
    def languages(self) -> List[Language]:
        """Distinct primary language objects."""
        seen: Set[str] = set()
        langs: List[Language] = []
        for v in self._videos:
            iso = v.primary_language.iso639_3
            if iso and iso not in seen:
                seen.add(iso)
                langs.append(v.primary_language)
        return langs

    @property
    def countries(self) -> List[str]:
        """Distinct country codes present in the collection."""
        seen: Set[str] = set()
        cc_list: List[str] = []
        for v in self._videos:
            cc = v.country_code
            if cc and cc not in seen:
                seen.add(cc)
                cc_list.append(cc)
        return cc_list

    @property
    def speakers(self) -> List[Speaker]:
        """All speakers present across videos."""
        result: List[Speaker] = []
        for v in self._videos:
            result.extend(v.speakers)
        return result

    @property
    def speaker_names(self) -> List[str]:
        """Distinct named speakers across the collection."""
        seen: Set[str] = set()
        names: List[str] = []
        for v in self._videos:
            for sp_name in v.speaker_names:
                if sp_name not in seen:
                    seen.add(sp_name)
                    names.append(sp_name)
        return names

    # -------------------------------------------------------------------------
    # Chainable Transformations & Navigation
    # -------------------------------------------------------------------------

    def first(self) -> Optional[Video]:
        """Return the first Video or None if empty."""
        return self._videos[0] if self._videos else None

    def last(self) -> Optional[Video]:
        """Return the last Video or None if empty."""
        return self._videos[-1] if self._videos else None

    def sample(self, k: int = 1, seed: Optional[int] = None) -> "VideoCollection":
        """Return a random sample of k videos."""
        if not self._videos:
            return VideoCollection([])
        k = min(k, len(self._videos))
        rng = random.Random(seed)
        sampled = rng.sample(self._videos, k)
        return VideoCollection(sampled)

    def filter(self, predicate: Callable[[Video], bool]) -> "VideoCollection":
        """Filter videos using an arbitrary predicate function."""
        return VideoCollection([v for v in self._videos if predicate(v)])

    def sort_by(
        self,
        key: Union[str, Callable[[Video], Any]],
        descending: bool = False,
    ) -> "VideoCollection":
        """
        Sort videos by a property name ('duration', 'upload_date', 'language_name', 'country')
        or a custom key function.
        """
        if callable(key):
            key_fn = key
        elif key in ("duration", "duration_seconds"):
            key_fn = lambda v: v.duration_seconds
        elif key in ("upload_date", "date"):
            key_fn = lambda v: v.upload_date
        elif key in ("language", "language_name"):
            key_fn = lambda v: v.primary_language.name.lower()
        elif key in ("iso", "iso639_3"):
            key_fn = lambda v: v.primary_language.iso639_3.lower()
        elif key in ("country", "country_code"):
            key_fn = lambda v: v.country_code or ""
        elif key in ("title",):
            key_fn = lambda v: v.title.lower()
        else:
            key_fn = lambda v: getattr(v, key, "")

        sorted_videos = sorted(self._videos, key=key_fn, reverse=descending)
        return VideoCollection(sorted_videos)

    def limit(self, n: int) -> "VideoCollection":
        """Return the first n videos."""
        return VideoCollection(self._videos[:n])

    def offset(self, n: int) -> "VideoCollection":
        """Skip the first n videos."""
        return VideoCollection(self._videos[n:])

    def page(self, page: int = 1, page_size: int = 20) -> "VideoCollection":
        """Return a 1-indexed page of videos."""
        if page < 1:
            page = 1
        start = (page - 1) * page_size
        end = start + page_size
        return VideoCollection(self._videos[start:end])

    def group_by(
        self,
        key: Union[str, Callable[[Video], Any]],
    ) -> Dict[Any, "VideoCollection"]:
        """Group videos by a key function or attribute name."""
        if callable(key):
            key_fn = key
        elif key in ("language", "iso", "iso639_3"):
            key_fn = lambda v: v.primary_language.iso639_3
        elif key in ("country", "country_code"):
            key_fn = lambda v: v.country_code or "Unknown"
        elif key in ("license",):
            key_fn = lambda v: v.license
        elif key in ("content_type",):
            key_fn = lambda v: v.content_type
        else:
            key_fn = lambda v: getattr(v, key, "Unknown")

        groups: Dict[Any, List[Video]] = {}
        for v in self._videos:
            group_key = key_fn(v)
            groups.setdefault(group_key, []).append(v)

        return {k: VideoCollection(vids) for k, vids in groups.items()}

    # -------------------------------------------------------------------------
    # Export & Serialization
    # -------------------------------------------------------------------------

    def to_dicts(self) -> List[Dict[str, Any]]:
        """Export collection to list of plain dictionaries."""
        return [v.to_dict() for v in self._videos]

    def to_json(self, indent: Optional[int] = 2) -> str:
        """Export collection to formatted JSON string."""
        return json.dumps(self.to_dicts(), ensure_ascii=False, indent=indent)

    def to_jsonl(self, filepath: Union[str, Path]) -> None:
        """Write collection to JSON Lines file."""
        target_path = Path(filepath)
        target_path.parent.mkdir(parents=True, exist_ok=True)
        with open(target_path, "w", encoding="utf-8") as f:
            for v in self._videos:
                f.write(v.to_json() + "\n")

    def to_dataframe(self):
        """
        Export collection to a pandas DataFrame if pandas is installed.
        Raises ImportError if pandas is not available.
        """
        try:
            import pandas as pd
        except ImportError:
            raise ImportError(
                "pandas is required for to_dataframe(). Install it with: pip install pandas"
            )

        flattened_rows = []
        for v in self._videos:
            flattened_rows.append({
                "id": v.id,
                "url": v.url,
                "title": v.title,
                "duration_seconds": v.duration_seconds,
                "duration_formatted": v.duration_formatted,
                "upload_date": v.upload_date,
                "license": v.license,
                "content_type": v.content_type,
                "iso639_3": v.primary_language.iso639_3,
                "bcp47": v.primary_language.bcp47,
                "language_name": v.primary_language.name,
                "glottocode": v.primary_language.glottocode,
                "autonym": v.primary_language.autonym,
                "dialect": v.primary_language.dialect,
                "country_code": v.country_code,
                "country_name": v.country_name,
                "has_subtitles": v.transcription.has_subtitles,
                "speakers": ", ".join(v.speaker_names),
            })
        return pd.DataFrame(flattened_rows)

File: src/collection.ts
/**
 * VideoCollection container for rich query results, aggregations, and transformations.
 */

import { Video, Language, Speaker } from './models';
import { VideoData } from './types';

// Simple seeded pseudo-random generator (Linear Congruential Generator / Mulberry32)
function createSeededRandom(seed: number): () => number {
  let s = seed >>> 0;
  return function () {
    s = (s + 0x6d2b79f5) >>> 0;
    let t = Math.imul(s ^ (s >>> 15), 1 | s);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export class VideoCollection implements Iterable<Video> {
  protected readonly _videos: Video[];

  constructor(videos: Video[] | readonly Video[] = []) {
    this._videos = [...videos];
  }

  public [Symbol.iterator](): Iterator<Video> {
    return this._videos[Symbol.iterator]();
  }

  public get length(): number {
    return this._videos.length;
  }

  public get(index: number): Video | undefined {
    return this._videos[index];
  }

  public at(index: number): Video | undefined {
    return this._videos.at(index);
  }

  public toArray(): Video[] {
    return [...this._videos];
  }

  public slice(start?: number, end?: number): VideoCollection {
    return new VideoCollection(this._videos.slice(start, end));
  }

  public includes(item: Video | string): boolean {
    if (typeof item === 'string') {
      return this._videos.some((v) => v.id === item);
    }
    return this._videos.some((v) => v.id === item.id);
  }

  // -------------------------------------------------------------------------
  // Aggregation & Summary Properties
  // -------------------------------------------------------------------------

  public get isEmpty(): boolean {
    return this._videos.length === 0;
  }

  public get ids(): string[] {
    return this._videos.map((v) => v.id);
  }

  public get urls(): string[] {
    return this._videos.map((v) => v.url);
  }

  public get embedUrls(): string[] {
    return this._videos.map((v) => v.embedUrl);
  }

  public get titles(): string[] {
    return this._videos.map((v) => v.title);
  }

  public get totalDurationSeconds(): number {
    return this._videos.reduce((acc, v) => acc + v.durationSeconds, 0);
  }

  public get totalDurationFormatted(): string {
    const total = this.totalDurationSeconds;
    const hours = Math.floor(total / 3600);
    const minutes = Math.floor((total % 3600) / 60);
    const seconds = total % 60;
    const parts: string[] = [];
    if (hours > 0) {
      parts.push(`${hours}h`);
    }
    if (minutes > 0 || hours > 0) {
      parts.push(`${minutes}m`);
    }
    parts.push(`${seconds}s`);
    return parts.join(' ');
  }

  public get averageDurationSeconds(): number {
    if (this._videos.length === 0) return 0;
    return this.totalDurationSeconds / this._videos.length;
  }

  public get isoCodes(): string[] {
    const seen = new Set<string>();
    const codes: string[] = [];
    for (const v of this._videos) {
      const code = v.primaryLanguage.iso639_3;
      if (code && !seen.has(code)) {
        seen.add(code);
        codes.push(code);
      }
    }
    return codes;
  }

  public get bcp47Codes(): string[] {
    const seen = new Set<string>();
    const codes: string[] = [];
    for (const v of this._videos) {
      const bcp = v.primaryLanguage.bcp47;
      if (bcp && !seen.has(bcp)) {
        seen.add(bcp);
        codes.push(bcp);
      }
    }
    return codes;
  }

  public get languageNames(): string[] {
    const seen = new Set<string>();
    const names: string[] = [];
    for (const v of this._videos) {
      const name = v.primaryLanguage.name;
      if (name && !seen.has(name)) {
        seen.add(name);
        names.push(name);
      }
    }
    return names;
  }

  public get languages(): Language[] {
    const seen = new Set<string>();
    const list: Language[] = [];
    for (const v of this._videos) {
      const iso = v.primaryLanguage.iso639_3;
      if (iso && !seen.has(iso)) {
        seen.add(iso);
        list.push(v.primaryLanguage);
      }
    }
    return list;
  }

  public get countries(): string[] {
    const seen = new Set<string>();
    const list: string[] = [];
    for (const v of this._videos) {
      const cc = v.countryCode;
      if (cc && !seen.has(cc)) {
        seen.add(cc);
        list.push(cc);
      }
    }
    return list;
  }

  public get speakers(): Speaker[] {
    const result: Speaker[] = [];
    for (const v of this._videos) {
      result.push(...v.speakers);
    }
    return result;
  }

  public get speakerNames(): string[] {
    const seen = new Set<string>();
    const list: string[] = [];
    for (const v of this._videos) {
      for (const spName of v.speakerNames) {
        if (!seen.has(spName)) {
          seen.add(spName);
          list.push(spName);
        }
      }
    }
    return list;
  }

  // -------------------------------------------------------------------------
  // Chainable Transformations & Navigation
  // -------------------------------------------------------------------------

  public first(): Video | null {
    return this._videos.length > 0 ? this._videos[0] : null;
  }

  public last(): Video | null {
    return this._videos.length > 0 ? this._videos[this._videos.length - 1] : null;
  }

  public sample(k: number = 1, seed?: number): VideoCollection {
    if (this._videos.length === 0) {
      return new VideoCollection([]);
    }
    const sampleSize = Math.min(k, this._videos.length);
    const rng = seed !== undefined ? createSeededRandom(seed) : Math.random;

    // Fisher-Yates shuffle on copy
    const pool = [...this._videos];
    for (let i = pool.length - 1; i > 0; i--) {
      const j = Math.floor(rng() * (i + 1));
      [pool[i], pool[j]] = [pool[j], pool[i]];
    }

    return new VideoCollection(pool.slice(0, sampleSize));
  }

  public filter(predicate: (video: Video, index: number) => boolean): VideoCollection {
    return new VideoCollection(this._videos.filter(predicate));
  }

  public sortBy(
    key: string | ((video: Video) => any),
    descending: boolean = false
  ): VideoCollection {
    let keyFn: (video: Video) => any;

    if (typeof key === 'function') {
      keyFn = key;
    } else if (key === 'duration' || key === 'duration_seconds') {
      keyFn = (v) => v.durationSeconds;
    } else if (key === 'upload_date' || key === 'date') {
      keyFn = (v) => v.uploadDate;
    } else if (key === 'language' || key === 'language_name') {
      keyFn = (v) => v.primaryLanguage.name.toLowerCase();
    } else if (key === 'iso' || key === 'iso639_3') {
      keyFn = (v) => v.primaryLanguage.iso639_3;
    } else if (key === 'country' || key === 'country_code') {
      keyFn = (v) => v.countryCode || '';
    } else if (key === 'title') {
      keyFn = (v) => v.title.toLowerCase();
    } else {
      keyFn = (v) => (v as any)[key] ?? '';
    }

    const sorted = [...this._videos].sort((a, b) => {
      const valA = keyFn(a);
      const valB = keyFn(b);
      if (valA < valB) return descending ? 1 : -1;
      if (valA > valB) return descending ? -1 : 1;
      return 0;
    });

    return new VideoCollection(sorted);
  }

  public limit(n: number): VideoCollection {
    return new VideoCollection(this._videos.slice(0, n));
  }

  public offset(n: number): VideoCollection {
    return new VideoCollection(this._videos.slice(n));
  }

  public page(page: number = 1, pageSize: number = 20): VideoCollection {
    const pageNum = Math.max(1, page);
    const start = (pageNum - 1) * pageSize;
    const end = start + pageSize;
    return new VideoCollection(this._videos.slice(start, end));
  }

  public groupBy(key: string | ((video: Video) => string)): Record<string, VideoCollection> {
    let keyFn: (video: Video) => string;

    if (typeof key === 'function') {
      keyFn = key;
    } else if (key === 'language' || key === 'iso' || key === 'iso639_3') {
      keyFn = (v) => v.primaryLanguage.iso639_3 || 'Unknown';
    } else if (key === 'country' || key === 'country_code') {
      keyFn = (v) => v.countryCode || 'Unknown';
    } else if (key === 'license') {
      keyFn = (v) => v.license || 'Unknown';
    } else if (key === 'content_type') {
      keyFn = (v) => v.contentType || 'Unknown';
    } else {
      keyFn = (v) => String((v as any)[key] || 'Unknown');
    }

    const groups: Record<string, Video[]> = {};
    for (const v of this._videos) {
      const gKey = keyFn(v);
      if (!groups[gKey]) {
        groups[gKey] = [];
      }
      groups[gKey].push(v);
    }

    const result: Record<string, VideoCollection> = {};
    for (const [k, vids] of Object.entries(groups)) {
      result[k] = new VideoCollection(vids);
    }
    return result;
  }

  // -------------------------------------------------------------------------
  // Export & Serialization
  // -------------------------------------------------------------------------

  public toDicts(): VideoData[] {
    return this._videos.map((v) => v.toDict());
  }

  public toJSON(indent: number = 2): string {
    return JSON.stringify(this.toDicts(), null, indent);
  }

  public toJSONL(): string {
    return this._videos.map((v) => v.toJSON()).join('\n') + (this._videos.length > 0 ? '\n' : '');
  }

  public toCSV(): string {
    const headers = [
      'id',
      'url',
      'title',
      'duration_seconds',
      'duration_formatted',
      'upload_date',
      'license',
      'content_type',
      'iso639_3',
      'iso_name',
      'bcp47',
      'glottocode',
      'glottolog_name',
      'glottolog_level',
      'wikitongues_classification',
      'wikitongues_lineage',
      'speaker_claim',
      'autonym',
      'country_code',
      'country_name',
      'has_subtitles',
      'speakers',
    ];

    const rows = this._videos.map((v) => {
      const escape = (val: any) => `"${String(val ?? '').replace(/"/g, '""')}"`;
      return [
        v.id,
        v.url,
        escape(v.title),
        v.durationSeconds,
        v.durationFormatted,
        v.uploadDate,
        v.license,
        v.contentType,
        v.primaryLanguage.iso639_3,
        escape(v.primaryLanguage.standards.iso639_3.name),
        v.primaryLanguage.bcp47,
        v.primaryLanguage.glottocode,
        escape(v.primaryLanguage.standards.glottolog.name),
        v.primaryLanguage.standards.glottolog.level,
        escape(v.primaryLanguage.wikitonguesClassification),
        escape(v.primaryLanguage.wikitonguesLineage ?? ''),
        escape(v.primaryLanguage.speakerClaim ?? ''),
        escape(v.primaryLanguage.autonym),
        v.countryCode || '',
        escape(v.countryName || ''),
        v.transcription.hasSubtitles,
        escape(v.speakerNames.join(', ')),
      ].join(',');
    });

    return [headers.join(','), ...rows].join('\n');
  }
}
